### PR TITLE
Feature: Add batch delete functionality for Environments 

### LINKED ISSUE #646: [FR] [RFC] vis.delete_env accepts regex pattern
In running experiments it is common, in developing/debugging stage, starting multiple runs. If each run starts an environment, the server will be soon flooded with too many envs. Currently, I usually use a combination of `get_env_list` and `delete_env` to clear all envs. But it would be cool to have a way to easily mass delete envs, e.g., by a regex pattern matching on the env names. :) 

It would be also cool to have a similar mass deleting functionality in UI as well.

### PR DESCRIPTION
**Pull request by vedansh-5**
## Description

**What changed:**

* Added checkbox-based multi-select UI in `EnvModal.js` for targeting multiple environments.
* Introduced "Delete Selected" button for bulk deletion.
* Updated state and API to handle batch deletion via environment ID arrays.
* Regenerated Cypress baselines for `modal.js` to reflect UI changes.

Fixes https://github.com/fossasia/visdom/issues/646

**Motivation and Context:**

* Resolves #646 by enabling mass deletion of environments, improving workflow efficiency for users handling many temporary environments.
* Updated Cypress baselines to align visual regression tests with the new modal layout.

## How Has This Been Tested?

* **Manual UI Testing:** Verified multi-select and bulk deletion flow; handled edge cases safely.
* **Automated Testing:** Ran `npm run test:init` and `npm run test:gui` to update snapshots and confirm tests pass.
* **Local Build:** Installed locally (`pip install -e .`) and confirmed successful build and runtime.

## Screenshots (if appropriate):
image

https://github.com/user-attachments/assets/7b598729-0e5f-4366-a153-a1d5d22c9b35

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor or cleanup (changes to existing code for improved readability or performance)

## Checklist:
- [ ] I adapted the version number under `py/visdom/VERSION` according to [[Semantic Versioning](https://semver.org/)](https://semver.org/)
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## Summary by Sourcery

Add support for selecting and deleting multiple environments at once from the Env modal and improve robustness of environment deletion on the server side.

New Features:
- Introduce checkbox-based multi-select UI in the Env modal to allow batch selection of environments for deletion.
- Add a Python client helper method to delete multiple environments in a single call.

Enhancements:
- Prevent errors when deleting environment JSON files by gracefully handling missing files and logging unexpected OS errors during deletion.
- Ensure active environments are deleted last in a batch to avoid disrupting the deletion flow.

Tests:
- Update Cypress modal tests to reflect the new environment deletion UI and interactions.

### CODE REVIEW

**Review by mariobehling — CHANGES_REQUESTED** [review]
Thanks! Please address AI comments and fix tests.

File: js/modals/EnvModal.js [hunk 1]
```diff
@@ -23,15 +23,20 @@ function EnvModal(props) {
 
   // change input / select value when activeEnv changes
   const [inputText, setInputText] = useState(activeEnv);
-  const [selectText, setSelectText] = useState(activeEnv);
+  const [selectText, setSelectText] = useState(activeEnv === 'main' ? [] : [activeEnv]);
   useEffect(() => {
     setInputText(activeEnv);
-    setSelectText(activeEnv);
+    setSelectText(activeEnv === 'main' ? [] : [activeEnv]);
   }, [activeEnv]);
```
[the remark this answers is not in the document]
Comment (vedansh-5) [reply js/modals/EnvModal.js]: changed in new commit

File: js/modals/EnvModal.js [hunk 2]
```diff
@@ -62,31 +67,62 @@ function EnvModal(props) {
...
-                {env}
-              </option>
-            );
-          })}
-        </select>
+     <div className="form-inline">
+        <div style={{ border: '1px solid #ccc', padding: '10px', height: '140px', overflowY: 'scroll', marginBottom: '10px', width: '100%', borderRadius: '4px', backgroundColor: '#fff' }}>
+          
+          <label style={{ display: 'block', fontWeight: 'bold', cursor: 'pointer' }}>
+            <input
```
[the remark this answers is not in the document]
Comment (vedansh-5) [reply js/modals/EnvModal.js:76]: follows all linting rules

File: py/visdom/__init__.py [hunk 3]
```diff
@@ -840,6 +840,17 @@ def close(self, win=None, env=None):
...
+    def delete_envs(self, env_list):
+        """This function deletes a list of environments."""
+        if not isinstance(env_list, list):
+            raise TypeError("env_list must be a list of strings")
+        responses = []
+        for env in env_list:
+            if not isinstance(env, str):
+                raise TypeError(f"Environment ID must be a string, got {type(env)}")
+            responses.append(self.delete_env(env))
+        return responses
```
[the remark this answers is not in the document]
Comment (vedansh-5) [reply py/visdom/__init__.py:872]: Implementing a true server-side batch deletion API is a larger architectural change. It requires adding a new delete_envs websocket endpoint to the backend, rewriting the frontend ApiProvider, and refactoring the socket handlers to safely batch the broadcast_envs state updates to avoid spamming the clients.

To keep the scope of this PR focused on the UI multi-select functionality, I'll follow this up after this PR is merged.

File: cypress/integration/modal.js [hunk 4]
```diff
@@ -57,8 +57,8 @@ describe('Test Env Modal', () => {
   it('Remove Env', () => {
     // delete fork
     cy.get(envbutton).click();
-    cy.get(envmodal + 'select').select(env + '_fork');
-    cy.contains('button', 'Delete').click();
+    cy.get(envmodal + 'input[type="text"]').type('_fork');
+    cy.contains('button', 'fork').click();
     cy.get(envmodal).type('{esc}');
```
[the remark this answers is not in the document]
Comment (vedansh-5) [reply cypress/integration/modal.js]: changed in latest push

**Comment by vedansh-5** [comment]
> Thanks! Please address AI comments and fix tests.

Hi @mariobehling 
I have addressed all the AI review comments in the latest commits, 
Regarding the failing tests: The `modal.js` Cypress test is currently failing because my PR changes the Environment Modal UI from a standard `<select>` to a checkbox-based multi-select list (I have added screenshots in the PR description). 
The test runner is still looking for the old HTML elements. I have updated the `modal.js` test suite locally to match the new UI (using `input[type="checkbox"]` and the "Delete Selected" button), so the tests should pass once these changes are merged, and the next CI run is triggered.

**Comment by Manik-Khajuria-5** [comment]
@vedansh-5 Is this a duplicate PR https://github.com/vedansh-5/visdom/pull/1 ?

**Comment by vedansh-5** [comment]
> @vedansh-5 Is this a duplicate PR [vedansh-5#1](https://github.com/vedansh-5/visdom/pull/1) ?

No @Manik-Khajuria-5, mentioned PR is in my fork, the current PR is on ```visdom/master```, The two are separate. Thanks!

**Comment by rajnisht7** [comment]
@vedansh-5 kindly address the failing tests

**Comment by vedansh-5** [comment]
> @vedansh-5 kindly address the failing tests

please go through the thread - the tests are failing because the UI is changed and new tests have been added, thanks!

**Comment by rajnisht7** [comment]
> > @vedansh-5 kindly address the failing tests
> 
> please go through the thread - the tests are failing because the UI is changed and new tests have been added, thanks!

as all the ci tests takes place on the basis of tests in forked repo branch not on the basis of main branch(of visdom), therefore failing tests clearly shows that the tests added are not proper or address the changes correctly
go through the tests properly and address them

**Comment by vedansh-5** [comment]
> > > @vedansh-5 kindly address the failing tests
> > 
> > 
> > please go through the thread - the tests are failing because the UI is changed and new tests have been added, thanks!
> 
> as all the ci tests takes place on the basis of tests in forked repo branch not on the basis of main branch(of visdom), therefore failing tests clearly shows that the tests added are not proper or address the changes correctly go through the tests properly and address them

ah alright, I'll fix these soon. thanks!

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@vedansh-5 Can u please fix failing test

**Comment by vedansh-5** [comment]
@Manik-Khajuria-5 @tonypzy @Jayantparashar10 @rajnisht7 Please review this PR

**Review by ArnavBallinCode — APPROVED** [review]
LGTM, tested works

**Comment by vedansh-5** [comment]
@Manik-Khajuria-5 @tonypzy @rajnisht7 @Jayantparashar10 
Please review this PR its ready to be merged and is a good functionality to be added. Thanks!

**Review by Jayantparashar10 — APPROVED** [review]

**Review by Manik-Khajuria-5 — APPROVED** [review]
@vedansh-5 Nice Work ! LGTM tested locally and read whole changes 

Image

**Review by tonypzy — APPROVED** [review]
Run locally and LGTM
Image

**Review by rajnisht7 — APPROVED** [review]

**Review by Saksham-Sirohi — APPROVED** [review]
